File: src/Models/CervicalCancer.js
import React, { useState } from 'react';
import "./Cc.css"
const CervicalCancerForm = () => {
 const [formData, setFormData] = useState({
    numberOfSexualPartners: '',
    firstSexualIntercourse: '',
    numOfPregnancies: '',
    smokes: '',
    smokesYears: '',
    smokesPacksYear: '',
    hormonalContraceptives: '',
    hormonalContraceptivesYears: '',
    iud: '',
    iudYears: '',
    stds: '',
    stdsNumber: '',
    stdsCondylomatosis: '',
    stdsCervicalCondylomatosis: '',
    stdsVaginalCondylomatosis: '',
    stdsVulvoPerinealCondylomatosis: '',
    stdsSyphilis: '',
    stdsPelvicInflammatoryDisease: '',
    stdsGenitalHerpes: '',
    stdsMolluscumContagiosum: '',
    stdsAIDS: '',
    stdsHIV: '',
    stdsHepatitisB: '',
    stdsHPV: '',
    stdsTimeSinceFirstDiagnosis: '',
    stdsTimeSinceLastDiagnosis: '',
    dxCancer: '',
    dxCIN: '',
    dxHPV: '',
    dx: '',
    hinselmann: '',
    citology: '',
    schiller: '',
 });

 const handleChange = (e) => {
    setFormData({ ...formData, [e.target.name]: e.target.value });
 };

 const handleSubmit = (e) => {
    e.preventDefault();
    // Here you would typically send the formData to your backend for processing
    console.log(formData);
 };

 return (
    <form onSubmit={handleSubmit}>
      <label>
        Number of sexual partners:
        <input type="number" name="numberOfSexualPartners" onChange={handleChange} />
      </label>
      <label>
        First sexual intercourse:
        <input type="number" name="firstSexualIntercourse" onChange={handleChange} />
      </label>
      <label>
        Number of pregnancies:
        <input type="number" name="numOfPregnancies" onChange={handleChange} />
      </label>
      <label>
        Smokes:
        <input type="number" name="smokes" onChange={handleChange} />
      </label>
      <label>
        Smokes (years):
        <input type="number" name="smokesYears" onChange={handleChange} />
      </label>
      <label>
        Smokes (packs/year):
        <input type="number" name="smokesPacksYear" onChange={handleChange} />
      </label>
      <label>
        Hormonal Contraceptives:
        <input type="number" name="hormonalContraceptives" onChange={handleChange} />
      </label>
      <label>
        Hormonal Contraceptives (years):
        <input type="number" name="hormonalContraceptivesYears" onChange={handleChange} />
      </label>
      <label>
        IUD:
        <input type="number" name="iud" onChange={handleChange} />
      </label>
      <label>
        IUD (years):
        <input type="number" name="iudYears" onChange={handleChange} />
      </label>
      <label>
        STDs:
        <input type="number" name="stds" onChange={handleChange} />
      </label>
      <label>
        STDs (number):
        <input type="number" name="stdsNumber" onChange={handleChange} />
      </label>
      <label>
        STDs: Condylomatosis:
        <input type="number" name="stdsCondylomatosis" onChange={handleChange} />
      </label>
      <label>
        STDs: Cervical Condylomatosis:
        <input type="number" name="stdsCervicalCondylomatosis" onChange={handleChange} />
      </label>
      <label>
        STDs: Vaginal Condylomatosis:
        <input type="number" name="stdsVaginalCondylomatosis" onChange={handleChange} />
      </label>
      <label>
        STDs: Vulvo-Perineal Condylomatosis:
        <input type="number" name="stdsVulvoPerinealCondylomatosis" onChange={handleChange} />
      </label>
      <label>
        STDs: Syphilis:
        <input type="number" name="stdsSyphilis" onChange={handleChange} />
      </label>
      <label>
        STDs: Pelvic Inflammatory Disease:
        <input type="number" name="stdsPelvicInflammatoryDisease" onChange={handleChange} />
      </label>
      <label>
        STDs: Genital Herpes:
        <input type="number" name="stdsGenitalHerpes" onChange={handleChange} />
      </label>
      <label>
        STDs: Molluscum Contagiosum:
        <input type="number" name="stdsMolluscumContagiosum" onChange={handleChange} />
      </label>
      <label>
        STDs: AIDS:
        <input type="number" name="stdsAIDS" onChange={handleChange} />
      </label>
      <label>
        STDs: HIV:
        <input type="number" name="stdsHIV" onChange={handleChange} />
      </label>
      <label>
        STDs: Hepatitis B:
        <input type="number" name="stdsHepatitisB" onChange={handleChange} />
      </label>
      <label>
        STDs: HPV:
        <input type="number" name="stdsHPV" onChange={handleChange} />
      </label>
      <label>
        STDs: Time since first diagnosis:
        <input type="number" name="stdsTimeSinceFirstDiagnosis" onChange={handleChange} />
      </label>
      <label>
        STDs: Time since last diagnosis:
        <input type="number" name="stdsTimeSinceLastDiagnosis" onChange={handleChange} />
      </label>
      <label>
        Dx: Cancer:
        <input type="number" name="dxCancer" onChange={handleChange} />
      </label>
      <label>
        Dx: CIN:
        <input type="number" name="dxCIN" onChange={handleChange} />
      </label>
      <label>
        Dx: HPV:
        <input type="number" name="dxHPV" onChange={handleChange} />
      </label>
      <label>
        Dx:
        <input type="number" name="dx" onChange={handleChange} />
      </label>
      <label>
        Hinselmann:
        <input type="number" name="hinselmann" onChange={handleChange} />
      </label>
      <label>
        Citology:
        <input type="number" name="citology" onChange={handleChange} />
      </label>
      <label>
        Schiller:
        <input type="number" name="schiller" onChange={handleChange} />
      </label>
      <button type="submit">Predict</button>
    </form>
 );
};

export default CervicalCancerForm;

File: src/Models/CervicalCancer.py
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix
file_path = r"src\Models\kag_risk_factors_cervical_cancer.csv"

data = pd.read_csv(file_path)
data.replace("?", np.nan, inplace=True)
for col in data.columns:
    if col in ['SomeNonNumericColumn1', 'SomeNonNumericColumn2']:
        continue
    data[col] = pd.to_numeric(data[col], errors='coerce')
data['Number of sexual partners'] = data['Number of sexual partners'].fillna(data['Number of sexual partners'].median())
data['First sexual intercourse'] = data['First sexual intercourse'].fillna(data['First sexual intercourse'].median())
data['Num of pregnancies'] = data['Num of pregnancies'].fillna(data['Num of pregnancies'].median())
data['Smokes'] = data['Smokes'].fillna(data['Smokes'].median())
data['Smokes (years)'] = data['Smokes (years)'].fillna(1)
data['Smokes (packs/year)'] = data['Smokes (packs/year)'].fillna(data['Smokes (packs/year)'].median())
data['Hormonal Contraceptives'] = data['Hormonal Contraceptives'].fillna(data['Hormonal Contraceptives'].median())
data['Hormonal Contraceptives (years)'] = data['Hormonal Contraceptives (years)'].fillna(data['Hormonal Contraceptives (years)'].median())
data['IUD'] = data['IUD'].fillna(0)
data['IUD (years)'] = data['IUD (years)'].fillna(0)
data['STDs'] = data['STDs'].fillna(1)
data['STDs (number)'] = data['STDs (number)'].fillna(data['STDs (number)'].median())
data['STDs:condylomatosis'] = data['STDs:condylomatosis'].fillna(data['STDs:condylomatosis'].median())
data['STDs:cervical condylomatosis'] = data['STDs:cervical condylomatosis'].fillna(data['STDs:cervical condylomatosis'].median())
data['STDs:vaginal condylomatosis'] = data['STDs:vaginal condylomatosis'].fillna(data['STDs:vaginal condylomatosis'].median())
data['STDs:vulvo-perineal condylomatosis'] = data['STDs:vulvo-perineal condylomatosis'].fillna(data['STDs:vulvo-perineal condylomatosis'].median())
data['STDs:syphilis'] = data['STDs:syphilis'].fillna(data['STDs:syphilis'].median())
data['STDs:pelvic inflammatory disease'] = data['STDs:pelvic inflammatory disease'].fillna(data['STDs:pelvic inflammatory disease'].median())
data['STDs:genital herpes'] = data['STDs:genital herpes'].fillna(data['STDs:genital herpes'].median())
data['STDs:molluscum contagiosum'] = data['STDs:molluscum contagiosum'].fillna(data['STDs:molluscum contagiosum'].median())
data['STDs:AIDS'] = data['STDs:AIDS'].fillna(data['STDs:AIDS'].median())
data['STDs:HIV'] = data['STDs:HIV'].fillna(data['STDs:HIV'].median())
data['STDs:Hepatitis B'] = data['STDs:Hepatitis B'].fillna(data['STDs:Hepatitis B'].median())
data['STDs:HPV'] = data['STDs:HPV'].fillna(data['STDs:HPV'].median())
data['STDs: Time since first diagnosis'] = data['STDs: Time since first diagnosis'].fillna(data['STDs: Time since first diagnosis'].median())
data['STDs: Time since last diagnosis'] = data['STDs: Time since last diagnosis'].fillna(data['STDs: Time since last diagnosis'].median())
data = pd.get_dummies(data = data, columns = ['Smokes','Hormonal Contraceptives','IUD','STDs',
                                      'Dx:Cancer','Dx:CIN','Dx:HPV','Dx','Hinselmann','Citology','Schiller'])

print(data.isnull().sum())
x = data.iloc[:,:46]
y = data.iloc[:,46]
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.4, random_state = 45)
mm = MinMaxScaler()
x_train = mm.fit_transform(x_train)
x_test = mm.fit_transform(x_test)

model = RandomForestClassifier()
model.fit(x_train, y_train)
y_pred = model.predict(x_test)
print(y_pred)
